daemon: Run once mode

ashcrow opened the pull request:
Non RHCOS nodes will need to apply an MC once and exit.

Requires: https://github.com/openshift/machine-config-operator/pull/139

/cc @aaronlevy @dustymabe @sdodson 



Still todo:
- [x] Rebase on #130 
- [x] Update remote http(s) `onceFrom` to request resources from the cluster
- [x] Update local ignition file `onceFrom` to be able to run without making requests for state from the cluster
- [x] Ensure kubernetes client doesn't attempt connection when there is no cluster

ashcrow commented:
/cc @sdemos @jlebon @kikisdeliveryservice for visibility as active fellow MCD devs.

ashcrow commented:
/cc @crawford 

@aaronlevy @sdodson

If we can pin down the format of the remote service I think we can actually write a card and execute on this. If at all possible I'd like the remote file to be the same as the ignition config ... OR the same as the CRD. The fewer things we have to parse to execute the same/similar operations the better IMHO.

aaronlevy commented:
I would lean toward the format being an ignition config (not CRD). My reasoning for this:

- We will need to generate raw/complete ignition configs for bootstrap/master/worker to support the eventual "bring your own RHCOS" environments. For example, "I already run a DC and have my own PXE infra - I just need the ignition profiles to use. 

- The bootstrap ignition config (largest/most complex) is already generated only as raw ignition and served from S3 - so we'd be wrapping it in a CRD just to immediately unwrap it anyway.

If we don't already generate ignition configs for master/worker (and instead just their CRD forms) - we likely should add them as an output format of the installer asset generation phase (cc @abhinavdahiya )

ashcrow commented:
@aaronlevy the MCD ends up consuming MCs (machine config) which wraps an ignition config with more metadata used for defining what version the host should be.

ashcrow commented:
> So would this seem reasonable to everyone:
>
>    - installer generates bootstrap machineConfig CR (in addition to raw ignition config)

:+1: 

>    - MCD can execute a run-once mode either from api endpoint (machineConfig CR) or locally (machineConfig CR)

That would sound reasonable to me. To reiterate to make sure I'm understanding properly, the api endpoint (full URI with scheme) would be passed to `--run-from` *or* a full file system path (/something/like/this/$FILENAME) would be passed to `--run-from`. In either case the same parsing mechanics used when watching CRD endpoints would be utilized, the only difference is it would parse and execute once, and then exit.

ashcrow commented:
If my reiteration above is correct I think we have enough to groom a card and do this work.

/cc @aaronlevy @sdodson @crawford

aaronlevy commented:
That sounds good to me. FWIW - naming of flags and such I have no strong feelings about

ashcrow commented:
OK cool. I've created a card based on this and we'll start filling it in soon.

ashcrow commented:
@sdemos When you have a few PTAL at what I have so far before I start wiring up a single process mode.

ashcrow commented on pkg/daemon/daemon.go:
```diff
@@ -102,6 +109,25 @@ func (dn *Daemon) Run(stop <-chan struct{}) error {
+		glog.V(2).Info("Running once per request")
+		// If we sense a local file has been provided ...
+		switch {
+		case strings.HasPrefix(string(os.PathSeparator), dn.onceFrom):
```
This does mean absolute. We could support relative ... I can add that in my next update.

sdemos replied:
either way is fine with me, but if it's just absolute we should make sure to be explicit about it, it might be a confusing error if you specify a relative file path and get told that you need to provide a file.

kikisdeliveryservice replied:
re: Errors, I'm wondering if we need to validate the URI before proceeding or if we think that the errors from the calls read files/pull configs will be enough?

ashcrow commented on cmd/machine-config-daemon/start.go:
```diff
@@ -32,6 +33,7 @@ func init() {
 	startCmd.PersistentFlags().StringVar(&startOpts.kubeconfig, "kubeconfig", "", "Kubeconfig file to access a remote cluster (testing only)")
 	startCmd.PersistentFlags().StringVar(&startOpts.nodeName, "node-name", "", "kubernetes node name daemon is managing.")
 	startCmd.PersistentFlags().StringVar(&startOpts.rootMount, "root-mount", "/rootfs", "where the nodes root filesystem is mounted for chroot and file manipulation.")
+	startCmd.PersistentFlags().StringVar(&startOpts.onceFrom, "once-from", "", "Runs the daemon once using the provided URI as it's configuration source")
```
Good catch. I'll clarify that this can be a path to a file or a URL. URI does make it sound like `http://`, `ftp://`, `file://`, etc.. would be supported when really it's `http[s]://` and `/...`.

kikisdeliveryservice commented on pkg/daemon/daemon.go:
```diff
@@ -370,6 +396,43 @@ func (dn *Daemon) Close() {
+// a MachineConfig struct.
+func (dn *Daemon) getMachineConfigFromFile(filePath string) (*mcfgv1.MachineConfig, error) {
+	config := mcfgv1.MachineConfig{}
+	data, err := ioutil.ReadFile(filePath)
```
Maybe we should add these ioutils to fsclient.go?:)

ashcrow replied:
Will do.

ashcrow replied:
https://github.com/openshift/machine-config-operator/pull/139

kikisdeliveryservice commented on cmd/machine-config-daemon/start.go:
```diff
@@ -32,6 +33,7 @@ func init() {
 	startCmd.PersistentFlags().StringVar(&startOpts.kubeconfig, "kubeconfig", "", "Kubeconfig file to access a remote cluster (testing only)")
 	startCmd.PersistentFlags().StringVar(&startOpts.nodeName, "node-name", "", "kubernetes node name daemon is managing.")
 	startCmd.PersistentFlags().StringVar(&startOpts.rootMount, "root-mount", "/rootfs", "where the nodes root filesystem is mounted for chroot and file manipulation.")
+	startCmd.PersistentFlags().StringVar(&startOpts.onceFrom, "once-from", "", "Runs the daemon once using the provided URI as it's configuration source")
```
minor typo: "its"

sdemos commented on pkg/daemon/daemon.go:
```diff
@@ -102,6 +109,25 @@ func (dn *Daemon) Run(stop <-chan struct{}) error {
+		}
+
+		// TODO: parse mcfgv1.MachineConfig{}
+		// TODO: run
```
yeah, makes sense. plus there is the work in #130 which is going to change that logic up even further. maybe we can just work on unifying it moving forward.

ashcrow replied:
@sdemos as in encapsulate the logic for `runOnce` on it's own as to make merging easy and follow up later to unify them?

sdemos replied:
yeah, since it would take some significant work to fully unify them today

ashcrow commented:
(not starting a retest yet as this requires another PR which hasn't merged yet due to yesterdays bot outage. It's being re-reviewed now)

ashcrow commented:
/retest

ashcrow commented:
Will rebase with #134 later today.

ashcrow commented on pkg/daemon/daemon.go:
```diff
@@ -129,6 +142,41 @@ func (dn *Daemon) Run(stop <-chan struct{}) error {
+
+	// At this point we have a populated MachineConfig struct for our desired config
+	// and we run update using an empty machineConfig as there is no provided current state.
+	oldConfig := mcfgv1.MachineConfig{}
```
@sdemos thoughts on this. Should we try to create an initial MC based on something, or does an empty one work?

sdemos replied:
an empty one will get rejected by `reconcilable` since there are "changes" to the config in places we can't fix.

it would be much better to use something that represents the actual system state. I'm not sure when or where this will be running exactly, will you be able to get it from kube? is it reasonable to expect the user to provide the current config? maybe we can have the current config always available somewhere on disk?

there are a few other possibilities if getting the current config isn't feasible. we could construct the old config from the irreconcilable parts of the new config. we could rip apart the update function so that the part we call always applies configs and we just say "trust us this one works". or we might be able to just pass it the new config for both the old and the new, since we already determined that we need to update, and most of the functionality in update just applies the config directly and doesn't diff. all of those approaches would break anything that does rely on the current configuration, like file deletion (I think that's the only thing right now, but conceivably there could be more in the future).

sdemos replied:
tl;dr is having an accurate current configuration seems like the vastly preferable option if that's at all possible, but you might be able to hack around it.

ashcrow replied:
I don't think we will be able to get the config from kubernetes but let me defer to folks who brought up and will be utilizing this use case. I tend to think we can't being that we can't pull the MC from the API itself, but it doesn't hurt to ask :smile: 

@sdodson / @aaronlevy at the point that MCD runs in the `onceFrom` path is the cluster up and serving machine configs?

aaronlevy replied:
In the case of the bootstrap ignition config, there would be no API/cluster yet (the output of this step would create one). And for bootstrap it's also not a state that we want to "reconcile". The bootstrap assets normally are on a separate throw-away node - but in the BYO case, they are run on one of the control-nodes - but then cleaned up after.

For all other cases of `onceFrom`, it would be retrieving the state from the API/cluster (either bootstrap API or real API).

sdemos replied:
so in the case of bootstrap we just want to blindly apply it, it sounds like. will we be required to delete files for the bootstrap case? at what point are we running on the bootstrap node? why can't we rely on ignition at that point, since if there is no bootstrap api then isn't this just the initial provisioning step that we are intervening in?

sorry if any of these questions are already answered somewhere else, I guess I haven't been keeping up with this feature as well as I thought I was.

aaronlevy replied:
> will we be required to delete files for the bootstrap case?

We do want to clean up these files after bootstrap completes, but I don't initially think it should be the responsibility of MCD (worry that there's fair amount of complexity here we could punt on a simple post-cleanup step).

> at what point are we running on the bootstrap node? 

In the BYO-RHEL host case, you've booted 3 masters and then we execute MCD with `--once-from=/on/disk/boostrap.ign` on one of them (no api is available yet). The bootstrap process then brings up a bootstrap MCO that will serve the master.ign. On all the master nodes then we would need to execute MCD `--once-from=https://control/plane` which would retrieve those roles. This overlap of MCD runs is where the deletion of bootstrap would get a little wonky (we shouldn't actually drop bootstrap assets until the control-plane is fully functional).

> why can't we rely on ignition at that point, since if there is no bootstrap api then isn't this just the initial provisioning step that we are intervening in?

This is to support RHEL nodes so they're already booted / no ignition involved. We're 'using' the ignition definitions to keep the source of truth for node configuration the same -- even if it's a bit of overloaded behavior.

ashcrow replied:
Thanks for the clarification @aaronlevy. We really appreciate it!

ashcrow commented:
Rebased.

Still todo:
- [ ] Update `process` to be able to run without making requests for state from the cluster
- [ ] Add tests

ashcrow commented:
Rebased on informers code by @yuqi-zhang

ashcrow commented:
Starting to work on:

- Update remote http(s) onceFrom to request resources from the cluster
- Update local ignition file onceFrom to be able to run without making requests for state from the cluster

ashcrow commented:
Added code and updated flow for:

- Update remote http(s) onceFrom to request resources from the cluster
- Update local ignition file onceFrom to be able to run without making requests for state from the cluster

It feels messy. As expected, we now have 3 cases for attempting updates instead of 1 and each has a slightly different path:

- Fully cluster integrated
- Partially cluster integrated
- Local like only

However, this is what I have so far. PTAL and help me simplify and restructure if possible!

sdemos commented:
the logic looks good to me. I toyed a bit with the code to see if I could think of a good way to simplify it, and none of my initial approaches panned out.

ashcrow commented:
/test e2e-aws

ashcrow commented:
/hold

ashcrow commented:
@aaronlevy @sdodson ready for review (though holding merge till I add tests)

ashcrow commented:
/test e2e-aws

ashcrow commented:
```
could not wait for pod to complete: could not wait for pod completion: the pod ci-op-i8c6txwz/e2e-aws failed after 13m46s (failed containers: setup, test):  unknown
```

and AWS errors.

ashcrow commented:
/test e2e-aws

ashcrow commented:
/test e2e-aws

ashcrow commented:
No code changes, but e2e now passes.

ashcrow commented:
> This is perhaps off topic for this PR but are there any suggestions for how we should run this? I assume we'll want to run it from a container rather than packaging it up and installing it on the host on BYOR. Should it be a static pod, should we craft a systemd unit, should we just run the container via ansible command module?

Good question. I'm not sure what the best way to kick off the container would be, but I agree that using a container would be the proper way to run MCD. I believe [this image](https://github.com/openshift/machine-config-operator/blob/master/Dockerfile.machine-config-daemon) could be used within or without the cluster.

ashcrow commented:
Assuming tests in CI all pass this should be ready to be tested. My quick smoke tests worked, but I'd like to have others give it a go.

yuqi-zhang commented on cmd/machine-config-daemon/start.go:
```diff
@@ -69,22 +67,44 @@ func runStartCmd(cmd *cobra.Command, args []string) {
+		if err != nil {
+			glog.Fatalf("failed to initialize daemon: %v", err)
+		}
+		err = dn.CheckStateOnBoot(stopCh)
```
So now that you start the informer in `NewClusterDrivenDaemon` above, we run into the problem that is: let's say a previous run degraded the node while updating, so we never got to the state where `desiredConfig` = `currentConfig`. Upon restarting, we start the informer and it will realize we need an update (dc != cc), and trigger that likely before we reach the CheckStateOnBoot (since in practice the update gets triggered quite often from each node). This way instead of sleeping (which preserves the actual degrade error from the previous run), we crash loop instead.

I recommend splitting out the starting of the informer to after this line instead.

ashcrow replied:
Updated with logic we came to via chat.

yuqi-zhang replied:
Confirmed, looks good!

yuqi-zhang commented on pkg/daemon/daemon.go:
```diff
@@ -177,38 +226,116 @@ func (dn *Daemon) CheckStateOnBoot(stop <-chan struct{}) error {
+		_, err := dn.prepUpdateFromCluster()
+		if err != nil {
+			glog.V(2).Infof("Unable to prep update: %s", err)
+			return
```
So for this and the following error, note that this is a callback function, so the return actual doesn't go anywhere useful. In that case, if say, the `prepUpdate` fails, we just return to... listening to the informer, which will trigger this again very soon, and it will just silently fail infinitely.

Barring a better way to do this, I recommend that on any update failure, for the time being to set the node to `degraded` and reboot the machine, so we can "catch" the error and sleep. I recognize this may not be the most clean way to handle this, and an alternative I tried was to throw/catch a `panic` and handle it that way if you think that makes sense.

ashcrow replied:
Hrm, that's a good point. `panic` handling is kind of a last ditch effort IMHO. I'll look at marking degraded and rebooting.

ashcrow replied:
Note that failure in `prepUpdateFromCluster()` will mark the node degraded on it's own. It doesn't, however, reboot. Will add that and some comments.

yuqi-zhang commented on pkg/daemon/daemon.go:
```diff
@@ -177,38 +226,116 @@ func (dn *Daemon) CheckStateOnBoot(stop <-chan struct{}) error {
+		}
+		if err = dn.executeUpdateFromCluster(); err != nil {
+			glog.V(2).Infof("Unable to execute update: %s", err)
+			return
```
same as above

ashcrow replied:
:+1:

ashcrow replied:
Re-checking the code `executeUpdateFromCluster()` will mark the node degraded and reboot upon error. I'll add some comments though.

yuqi-zhang commented on pkg/daemon/daemon.go:
```diff
@@ -177,38 +226,116 @@ func (dn *Daemon) CheckStateOnBoot(stop <-chan struct{}) error {
+			glog.V(2).Infof("Unable to prep update: %s", err)
+			return
+		}
+		if err = dn.executeUpdateFromCluster(); err != nil {
```
If there is no update from the above `prepUpdateFromCluster`, shouldn't we not run this? Currently we only return if that previous function errors.

Also, we should be passing the `desiredConfig` from the `prepUpdate` above, otherwise we'll hit https://github.com/openshift/machine-config-operator/pull/145 again, right?

ashcrow replied:
Agreed on first point.

On the second point are you recommending that we should return the `desiredConfig` as a return value for use later?

yuqi-zhang replied:
Actually maybe I mis-understood the update method the first time around, I think this should be fine.

Originally, the main concern being that let's say 2 updates happen in quick succession, then the first update is not stored, and we might see the second `desiredConfig` when we poll it again at: https://github.com/openshift/machine-config-operator/pull/126/files#diff-06b532a2a5a5d3ad1d5341b3e00945efR404 , but that should be fine since we'd just apply 2 updates and bring it to the most updated "desired state"? Correct me if I'm wrong in this understanding. If that's how it operates, then it's fine to leave as is.

ashcrow replied:
`prepUpdateFromCluster` is:

yuqi-zhang replied:
Yes, I mistook the application of config and updating the annotation. This should be fine.

ashcrow commented:
/test e2e-aws

Router didn't create.

vrutkovs commented:
Looks good to me. MCD would probably have to run in a plain container as its ready to use, while kubelet needs to be pre-configured on the node

ashcrow commented:
Thanks @vrutkovs! With the `--once-from` the kubelet config would either need to be part of the Machine Config file provided during the single run, OR, if on a non-immutable host, configured by an external entity as noted.

ashcrow commented:
/test e2e-aws

Router didn't start again.

ashcrow commented:
With the updates from the last review I'm getting an error when running once from a file only:

```
I1107 15:50:24.160249    1292 update.go:81] Checking if configs are reconcilable                                                                  
W1107 15:50:24.160256    1292 update.go:91] daemon can't reconcile state!                                                                         
W1107 15:50:24.160273    1292 update.go:92] Ignition version mismatch between old and new config: old:  new: 2.2.0                                
F1107 15:50:24.160279    1292 start.go:128] failed to run: daemon can't reconcile this config     
```

Looking into this.

ashcrow commented:
`daemon/update: Allow reconcile skip` assumes there was no previous state when we are in runOnce mode and the previous MC has no Kind.

Tested with:

```
./machine-config-daemon start -v 2 --node-name=test --root-mount=/ --once-from=/tmp/inside.ign
```
Using a valid MC at `/tmp/inside.ign`

@yuqi-zhang :arrow_up:

kikisdeliveryservice commented:
@ashcrow agree that you will def have to skip using func reconcilable() for this PR. :+1:

yuqi-zhang commented on cmd/machine-config-daemon/start.go:
```diff
@@ -69,22 +67,47 @@ func runStartCmd(cmd *cobra.Command, args []string) {
+		if err != nil {
+			glog.Fatalf("failed to initialize daemon: %v", err)
+		}
+		err = dn.CheckStateOnBoot(stopCh)
```
Another note for here: I remember there were issues doing state checking, etc. before the `Chroot` and the `Chdir` right after, I will double check.

yuqi-zhang commented on pkg/daemon/daemon.go:
```diff
@@ -67,59 +71,112 @@ const (
+
+	ctx := common.CreateControllerContext(cb, stopCh, componentName)
+
+	ctx.KubeInformerFactory.Start(ctx.Stop)
```
In the current workflow we do this after setting up the nodeInformer/Lister. I'm not sure if the ordering will cause issues, but I remember that is what we did elsewhere for MCO/MCC, so perhaps move this line and the following line to the end of this function?

ashcrow replied:
Updated and moved start to the end of this function before return.

yuqi-zhang commented on pkg/daemon/daemon.go:
```diff
@@ -177,38 +234,123 @@ func (dn *Daemon) CheckStateOnBoot(stop <-chan struct{}) error {
+		}
+		// At this point we have verified we need to update
+		if err := dn.executeUpdateFromClusterWithMachineConfig(machineConfig); err != nil {
+			glog.Warningf("Unable to update: %s", err)
```
We do already `Errorf` inside the `executeUpdateFromClusterWithMachineConfig`, probably unnecessary

yuqi-zhang commented on pkg/daemon/daemon.go:
```diff
@@ -177,38 +234,123 @@ func (dn *Daemon) CheckStateOnBoot(stop <-chan struct{}) error {
+			// Note that if executeUpdateFromCluster errors it will mark the node
+			// degraded and reboot.
+			if err = dn.executeUpdateFromCluster(); err != nil {
+				glog.V(2).Infof("Unable to execute update: %s", err)
```
Can probably remove, errors are logged in the update execution

yuqi-zhang commented on pkg/daemon/daemon.go:
```diff
@@ -177,38 +234,123 @@ func (dn *Daemon) CheckStateOnBoot(stop <-chan struct{}) error {
+	if node.Annotations[DesiredMachineConfigAnnotationKey] == node.Annotations[CurrentMachineConfigAnnotationKey] {
 		// No actual update to the config
-		return
+		glog.V(2).Info("No updating is required")
```
I suggest removing this since this will likely get logged a lot in cluster operation.

ashcrow commented:
Same router flake. Waiting till after manual testing before trying an e2e test.

abhinavdahiya commented:
@ashcrow
```console
Setup success
NAME                           STATUS                     ROLES     AGE       VERSION
ip-10-0-130-104.ec2.internal   Ready,SchedulingDisabled   worker    25m       v1.11.0+d4cacc0
ip-10-0-154-42.ec2.internal    Ready,SchedulingDisabled   worker    25m       v1.11.0+d4cacc0
ip-10-0-166-190.ec2.internal   Ready,SchedulingDisabled   worker    25m       v1.11.0+d4cacc0
ip-10-0-27-206.ec2.internal    Ready,SchedulingDisabled   master    30m       v1.11.0+d4cacc0
ip-10-0-32-194.ec2.internal    Ready,SchedulingDisabled   master    29m       v1.11.0+d4cacc0
ip-10-0-7-73.ec2.internal      Ready,SchedulingDisabled   master    30m       v1.11.0+d4cacc0
```

this is not a router flake.. somebody is making the node SchedulingDisabled,
MCD does cordon nodes... might be something to look into...

ashcrow commented:
@abhinavdahiya hrm ok. Thanks for the note. You're right, MCD does set the `degraded` label so it's possible that it's flowing back into disabled. @yuqi-zhang is testing it out manually now. If he sees the same issue we'll have data to debug with.

yuqi-zhang commented:
@ashcrow Locally, it seems that the daemons are running fine (no sign of degradation), but no updates are being applied when I do `oc apply` (i.e. there are no reboots). I suspect this could be because of the ordering of informer workflow as mentioned above? Not sure what is causing the e2e degradation though.

ashcrow commented:
Updated the informer code per request.

kikisdeliveryservice commented:
Seems like another flake? :(

abhinavdahiya commented:
> Seems like another flake? :(

We are not seeing this specific error anywhere else **yet**
```console
NAME                           STATUS                     ROLES     AGE       VERSION
ip-10-0-12-35.ec2.internal     Ready,SchedulingDisabled   master    1h        v1.11.0+d4cacc0
ip-10-0-137-214.ec2.internal   Ready,SchedulingDisabled   worker    1h        v1.11.0+d4cacc0
ip-10-0-145-104.ec2.internal   Ready,SchedulingDisabled   worker    1h        v1.11.0+d4cacc0
ip-10-0-175-235.ec2.internal   Ready,SchedulingDisabled   worker    1h        v1.11.0+d4cacc0
ip-10-0-23-254.ec2.internal    Ready,SchedulingDisabled   master    1h        v1.11.0+d4cacc0
ip-10-0-44-102.ec2.internal    Ready,SchedulingDisabled   master    1h        v1.11.0+d4cacc0
Waiting for router to be created ...
```

That makes it seem like it is not a flake.

ashcrow commented:
Agreed, I don't think this is a flake. It does look like a flake we hit previously BUT @yuqi-zhang and I have found cases where nodes are set degraded when they shouldn't be. We're working on this now.

ashcrow commented:
/hold cancel

- e2e now passes
- onceFrom a local file verified to work

kikisdeliveryservice commented on pkg/daemon/daemon.go:
```diff
@@ -120,6 +151,12 @@ func New(
 func (dn *Daemon) Run(stop <-chan struct{}) error {
+	// Catch quickly if we've been asked to run once.
+	if dn.onceFrom != "" {
+		glog.V(2).Info("Running once per request")
```
Is this logging clear enough? Would "daemon running once per request" be clearer? Thoughts?

ashcrow replied:
I like your wording better. How about I rework some of the log strings in a follow up?

kikisdeliveryservice replied:
Sounds great!

kikisdeliveryservice commented:
Also, thanks for adding comments in this PR @ashcrow , it makes it very easy to read through! :+1:

yuqi-zhang commented:
Some manual testing passes for me both for cluster operation and runOnce with qemu. Will LGTM after logs are cleaned up.

Thanks for the work @ashcrow !

yuqi-zhang commented:
/lgtm

ashcrow commented:
Thanks @yuqi-zhang and @kikisdeliveryservice! I'll do the log clean up in another PR post merge.
